### PR TITLE
Configure GitHub Actions to run Super-Linter

### PR DESCRIPTION
**Pull request by lparsons**
This Pull Request sets up a GitHub Actions workflow that runs the GitHub Super-Linter on the entire codebase. It also formats the files according to the standards specified by the linters.

Super-Linter is a collection of `bash` scripts that run various "standard" linters for many different types of files. The hope is that this helps to:

* Prevent broken code from being uploaded to the default branch (main)
* Help establish coding best practices across multiple languages
* Build guidelines for code layout and format
* Automate the process to help streamline code reviews

The linters that are in use right now are:

* [Markdown-lint](https://github.com/igorshubovych/markdownlint-cli#readme) - `.markdown-lint.yml`
* [Flake8](https://flake8.pycqa.org/en/latest/) - `.flake8`
* [Pylint](https://www.pylint.org/) - `.python-lint` -> `.pylintrc`
* [Black](https://black.readthedocs.io/en/stable/) - `.python-black` - Note that `black` will actually reformat your code for you, so just run `black .` before committing to fix everything up.
* [isort](https://pycqa.github.io/isort/) - `.isort.cfg` - `isort` also automatically formats files for you (`isort .`)

The python linters can all be installed via `pip`. To make it simpler to install them on a development machine, this PR creates a `requirements/dev.txt` file that can be used to create a virtual environment with the linters installed in addition to the core dependencies. Using `requirements.txt` or `requirements/prod.txt` will install only those things needed for production.

Finally, the configuration for the linters is borrowed from Super-Linter which attempts to provide some "best-practices" and also help ensure the linters all get along. There are two changes to the defaults:

* `.isort.cfg`
    * Added config to include our Django apps as first party packages. For some reason running under Super-Linter didn't do this in GitHub actions.
* `.flake8`
    * ignore E231 There should be whitespace after the characters ,, ;, and :. black; fixed in black 20 (https://github.com/psf/black/issues/1289)

### CODE REVIEW

File: .isort.cfg [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[settings]
+profile=
+
+; vertical hanging indent mode also used in black configuration
+multi_line_output = 3
+
+; necessary because black expect the trailing comma
```
Comment (hepcat72) [inline .isort.cfg:7]:
```suggestion
; necessary because black expects the trailing comma
```

File: .github/workflows/superlinter.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+name: Super-Linter
+
+# Run this workflow every time a new commit pushed to your repository
```
Comment (hepcat72) [inline .github/workflows/superlinter.yml:3]:
```suggestion
# Run this workflow every time a new commit is pushed to your repository
```

File: .markdown-lint.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#
+
+###############
+# Rules by id #
+###############
+MD004: false                  # Unordered list style
+MD007:
+  indent: 2                   # Unordered list indentation
+MD013:
+  line_length: 400            # Line length 80 is far to short
```
Comment (hepcat72) [inline .markdown-lint.yml:25]:
```suggestion
  line_length: 400            # Line length 80 is far too short
```

File: .markdown-lint.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+MD026:
+  punctuation: ".,;:!。，；:"  # List of not allowed
```
Comment (hepcat72) [inline .markdown-lint.yml:27]: What's the story behind the repeated weird looking punctuation?

File: .markdown-lint.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+MD029: false                  # Ordered list item prefix
+MD033: false                  # Allow inline HTML
+MD036: false                  # Emphasis used instead of a heading
+
+#################
+# Rules by tags #
+#################
+blank_lines: false  # Error on blank lines
```
Comment (hepcat72) [inline .markdown-lint.yml:35]: Why error on blank lines?  I think it makes it easier to look at the raw file.

File: .python-lint [hunk 6]
```diff
@@ -0,0 +1,544 @@
+[MASTER]
```
Comment (hepcat72) [inline .python-lint:1]: I can't help but notice that the max line length in this file is 80.  Why does that differ from the line length for python specified for flake8?  I mean, I know this was a default file.  I'm just a perfectionist about consistency I guess.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -89,3 +99,18 @@ Verify you can run the development server.  Run:
 Then go to this site in your web browser:
 
     http://127.0.0.1:8000/
+
+## Code Formatting Standards
+
+All Pull Requests must pass linting prior to being merged.
+
+Currently, linting is using by [GitHub's
```
Comment (hepcat72) [inline CONTRIBUTING.md:107]:
```suggestion
Currently, all pushed are linted using [GitHub's
```

File: DataRepo/admin.py [hunk 8]
```diff
@@ -1,3 +1,3 @@
-from django.contrib import admin
+# from django.contrib import admin
```
Comment (hepcat72) [inline DataRepo/admin.py:1]: Commented line.  If it's not needed, I say remove.

File: DataRepo/models.py [hunk 9]
```diff
@@ -1,3 +1,3 @@
-from django.db import models
+# from django.db import models
```
Comment (hepcat72) [inline DataRepo/models.py:1]: Commented line.  If it's not needed, I say remove.

File: DataRepo/tests.py [hunk 10]
```diff
@@ -1,3 +1,3 @@
-from django.test import TestCase
+# from django.test import TestCase
```
Comment (hepcat72) [inline DataRepo/tests.py:1]: Commented line.  If it's not needed, I say remove.

File: DataRepo/views.py [hunk 11]
```diff
@@ -1,10 +1,11 @@
-from django.shortcuts import render
-
 from django.http import HttpResponse
 
+# from django.shortcuts import render
```
Comment (hepcat72) [inline DataRepo/views.py:3]: Commented line.  If it's not needed, I say remove.

File: requirements/dev.txt [hunk 12]
```diff
@@ -0,0 +1,7 @@
+# Specifies only dev-specific requirements
+# But imports the common ones too
+-r common.txt
```
Comment (hepcat72) [inline requirements/dev.txt:3]: Are you sure we don't need the requirements directory here?  It's supplied in the requirements.txt file.

Comment (hepcat72) [reply requirements/dev.txt:3]: I mean, I assume this works, but I just thought I might learn something by asking.

**Review by hepcat72 — APPROVED** [review]
Nothing blocking.

File: .flake8 [hunk 13]
```diff
@@ -0,0 +1,4 @@
+[flake8]
+max-line-length = 120
```
[the remark this answers is not in the document]
Comment (lparsons) [reply .flake8:2]: Yeah, this looked weird to me to, but it's just something to keep `flake8` from complaining. `black` does the formatting and [keeps all lines at 88 max](https://black.readthedocs.io/en/stable/the_black_code_style.html#line-length). It's probably easiest to just quickly run `black` before each commit. I'm actually thinking I might try setting up some [pre-commit hooks](https://githooks.com/). BTW, all of these settings are defaults from Super-Linter (with just a couple of exceptions noted in in the PR comments).

File: .github/workflows/superlinter.yml [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+on: push
+
+jobs:
+  # Set the job key. The key is displayed as the job name
+  # when a job name is not provided
+  super-lint:
+    # Name the Job
+    name: Lint code base
+    # Set the type of machine to run on
+    runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (lparsons) [reply .github/workflows/superlinter.yml:13]: I agree in principal. In this case I don't think GitHub actions machines change very often. And this is for the linters, not running any specific code. I do think the linters could change on us a bit too, which is annoying, but generally a single commit to fix any changes (possibly using `--ignore-rev`) will fix resolve that.

File: .markdown-lint.yml (hunk 4 above)
Comment (lparsons) [reply .markdown-lint.yml:27]: I'm just going with Super-Linter defaults here: https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md026

File: .markdown-lint.yml (hunk 3 above)
Comment (lparsons) [reply .markdown-lint.yml:25]: I'll update, though this is a copy from the Super-Linter defaults (to make customizing or locally running easier).

File: .markdown-lint.yml (hunk 5 above)
Comment (lparsons) [reply .markdown-lint.yml:35]: Leaving the defaults for now.

File: DataRepo/admin.py (hunk 8 above)
Comment (lparsons) [reply DataRepo/admin.py:1]: This is a special case that I commented only to keep the pylint from complaining about unusued imports. I wanted to keep the stub info in there.

File: requirements/dev.txt (hunk 12 above)
Comment (lparsons) [reply requirements/dev.txt:3]: Good catch, but this does work. It seems that the reference is relative to the requirements file.

**Comment by hepcat72** [comment]
Resolved all comments. Full speed ahead!

**Pull request merged**